app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java

```diff
@@ -493,15 +534,72 @@ private void queueDownload(Download download) {
+            query.setFilterById(mDownloadReference);
+            try (Cursor cursor = manager.query(query)) {
+                if (cursor.moveToFirst()) {
+                    int columnIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+                    if (DownloadManager.STATUS_SUCCESSFUL == cursor.getInt(columnIndex)) {
+                        String uriString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
+
+                        if (uriString.startsWith("file://")) {
+                            uriString = uriString.substring(7);
+                        }
+                        final String localUri = uriString;
+
+                        final String fileExtension = MimeTypeMap.getFileExtensionFromUrl(localUri);
+                        final String fileName = URLUtil.guessFileName(localUri, null, null);
+                        final String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
+
+                        final File file = new File(localUri);
+                        final Uri uriFile = FileProvider.getUriForFile(getContext(), BuildConfig.APPLICATION_ID + ".fileprovider", file);
+
+                        final Intent openFileIntent = new Intent(Intent.ACTION_VIEW);
+                        openFileIntent.setDataAndType(uriFile, mimeType);
+                        openFileIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+                        final PackageManager packageManager = getContext().getPackageManager();
+                        final List<ResolveInfo> resolveInfoList = packageManager.queryIntentActivities(openFileIntent, 0);
+                        final boolean activitiesFoundForIntent = (resolveInfoList.size() > 0);
+
+                        final Snackbar snackbar = Snackbar
+                                .make(browserContainer, String.format(getString(R.string.download_snackbar_finished), fileName), Snackbar.LENGTH_LONG)
+                                .setAction(getString(R.string.download_snackbar_open), new View.OnClickListener() {
+                                    @Override
+                                    public void onClick(View view) {
+                                        if (activitiesFoundForIntent && mimeType != null) {
+                                            startActivity(openFileIntent);
+                                        }
+                                    }
+                                });
+                        if (activitiesFoundForIntent) {
+                            snackbar.setActionTextColor(ContextCompat.getColor(getContext(), R.color.download_dialog_text));
+                        } else {
+                            snackbar.setAction("", null);
+                        }
+                        snackbar.show();
+                        getActivity().unregisterReceiver(downloadBroadcastReceiver);
```

Comment:
We still need to get here in order to successfully unregister (e.g. we do not if there's an exception or the query is empty). If we do not then this receiver (and via reference this fragment) will exist indefinitely in memory. Register/Unregister in onResume()/onPause() is safer!